### PR TITLE
[DOC] StringIO doc enhancements

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Treated:
- ::new
- ::open
- #string
- #string=
- #close
- #close_read
- #close_write
- #closed?
- #closed_read?
- #closed_write?
- #eof?

### CODE REVIEW

**Comment by BurdetteLamar** [comment]
@nobu, I will merge this only after requesting your approval.  Meanwhild, @peterzhu2118 and I can thrash out details?

File: ext/stringio/stringio.c
Comment (peterzhu2118) [inline ext/stringio/stringio.c:535]: Might be interesting to mention that the cursor is set to 0.

Reading:

```ruby
irb(main):013:0> io = StringIO.new("hello")
=> #<StringIO:0x0000000108cbd208>
irb(main):014:0> io.read
=> "hello"
irb(main):015:0> io.pos
=> 5
irb(main):016:0> io.string = "foo"
=> "foo"
irb(main):017:0> io.pos
=> 0
irb(main):018:0> io.read
=> "foo"
```

Writing:
```ruby
irb(main):034:0> io = StringIO.new("hello")
=> #<StringIO:0x0000000108d4e280>
irb(main):035:0> io.string = "hello world 123"
=> "hello world 123"
irb(main):036:0> io.write("foo")
=> 3
irb(main):037:0> io.string
=> "foolo world 123"
```

File: ext/stringio/stringio.c
Comment (peterzhu2118) [inline ext/stringio/stringio.c:569]: I think it would be useful to keep the part about raising `IOError` when attempting to read from or write to a closed StringIO.

File: ext/stringio/stringio.c
Comment (peterzhu2118) [inline ext/stringio/stringio.c:571]:
```suggestion
 * Related: StringIO#close_read, StringIO#close_write.
```

File: ext/stringio/stringio.c
Comment (peterzhu2118) [inline ext/stringio/stringio.c:587]:
```suggestion
 * Related: StringIO#close, StringIO#close_write.
```

File: ext/stringio/stringio.c
Comment (peterzhu2118) [inline ext/stringio/stringio.c:606]:
```suggestion
* Related: StringIO#close, StringIO#close_read.
```

File: ext/stringio/stringio.c
Comment (peterzhu2118) [inline ext/stringio/stringio.c:674]: Minor punctuation nit:

```suggestion
  * Returns +true+ if positioned at end-of-stream; +false+ otherwise;
```

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c]: Added.

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c:569]: Added.

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c]: Fixed.

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c]: Fixed.

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c]: Fixed.

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c]: Added comma (consistent w/others).

File: ext/stringio/stringio.c
Comment (nobu) [inline ext/stringio/stringio.c:407]: `mode` is defaulted to `r` if `string` is frozen.
Is this worth to describe?

File: ext/stringio/stringio.c
Comment (nobu) [inline ext/stringio/stringio.c:533]: Isn't this confusing?

File: ext/stringio/stringio.c
Comment (nobu) [inline ext/stringio/stringio.c:622]: Isn't this sentence needed?

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c:411]: Added to ::new and ::open..

File: ext/stringio/stringio.c
Comment (BurdetteLamar) [reply ext/stringio/stringio.c:622]: Added (three places).

**Review by peterzhu2118 — APPROVED** [review]
Looks good!

**Comment by BurdetteLamar** [comment]
@nobu, I don't want to merge without your approval.

**Comment by BurdetteLamar** [comment]
@nobu, hoping for your approval.  There's more I can do here once this is merged.

**Comment by BurdetteLamar** [comment]
@nobu, you're the maintainer here, right?  Any further review?

**Comment by BurdetteLamar** [comment]
@nobu, let me know if you have further comments.

**Pull request merged**
